src/transformers/models/bloom/modeling_bloom.py

```diff
@@ -870,16 +871,194 @@ def _reorder_cache(past: Tuple[Tuple[torch.Tensor]], beam_idx: torch.Tensor) ->
-    `pad_token_id` is defined in the configuration, it finds the last token that is not a padding token in each row. If
-    no `pad_token_id` is defined, it simply takes the last value in each row of the batch. Since it cannot guess the
-    padding tokens when `inputs_embeds` are passed instead of `input_ids`, it does the same (take the last value in
-    each row of the batch).
+    def set_output_embeddings(self, new_embeddings):
+        self.lm_head = new_embeddings
+
+    def prepare_inputs_for_generation(self, input_ids, past=None, **kwargs):
+        # only last token for inputs_ids if past is defined in kwargs
+        if past:
+            input_ids = input_ids[:, -1].unsqueeze(-1)
+
+        attention_mask = kwargs.get("attention_mask", None)
+        position_ids = kwargs.get("position_ids", None)
+        prefix_length = kwargs.get("prefix_length", None)
+
+        causal_mask = kwargs.get("causal_mask", None)
+        if past is None:
+            # causal_mask is a class attribute so that the prefix mask
+            # set at the start of generation persists throughout timesteps.
+            # here, we reset it at start of a generation call so that forward() creates a new one.
+            self.causal_mask = causal_mask
+
+        if attention_mask is not None and position_ids is None:
+            # create position_ids on the fly for batch generation
+            position_ids = attention_mask.long().cumsum(-1) - 1
+            position_ids.masked_fill_(attention_mask == 0, 1)
+            if past:
+                position_ids = position_ids[:, -1].unsqueeze(-1)
+        else:
+            position_ids = None
+
+        # determine length of prefix, if none is provided
+        if prefix_length is None and self.causal_mask is None and past is None:
+            if attention_mask is not None:
+                # TODO: what will attn mask look like for concatenated inputs+targets, in a decoder?
+                prefix_length = attention_mask.sum(-1)
+            else:
+                prefix_length = torch.LongTensor([input_ids.shape[-1] for _ in range(input_ids.shape[0])])
+
+        return {
+            "input_ids": input_ids,
+            "past_key_values": past,
+            "use_cache": kwargs.get("use_cache"),
+            "position_ids": position_ids,
+            "attention_mask": attention_mask,
+            "prefix_length": prefix_length,
+            "causal_mask": self.causal_mask,
+        }
+
+    @add_start_docstrings_to_model_forward(BLOOM_INPUTS_DOCSTRING)
+    @add_code_sample_docstrings(
+        processor_class=_TOKENIZER_FOR_DOC,
+        checkpoint=_CHECKPOINT_FOR_DOC,
+        output_type=CausalLMOutputWithCrossAttentions,
+        config_class=_CONFIG_FOR_DOC,
+    )
+    def forward(
+        self,
+        input_ids=None,
+        past_key_values=None,
+        attention_mask=None,
+        causal_mask=None,
```

Comment:
We need a nice docstring for `causal_mask` here - also do we need a whole new model class for this? Can't we just add a new `causal_mask` argument to the `...ForCausalLM` model?